rfcs/20230801-max-partition-policy/README.md

```diff
@@ -0,0 +1,116 @@
+# Introducing Max Partition Policy to oneDNN Graph API
+
+## Motivation
+
+oneDNN Graph Compiler is currently a experimental backend for oneDNN Graph API.
+It provides
+optimized implementations for complex computational graphs including multi-head
+attention (MHA), multi-layer perceptron (MLP), and convolution residual blocks
+over typical data types for both inference and training.
+
+To further boost performance and expose more aggressive optimizations to users,
+oneDNN Graph Compiler is requesting to add max partition
+policy to oneDNN Graph API.
+
+## Current Status
+
+In oneDNN Graph API programming model, a computation graph is passed to library
+and then optimized partitions are returned by the library by
+`graph.get_partitions()` API.
+
+```c
+dnnl_status_t dnnl_graph_graph_filter(
+        dnnl_graph_graph_t graph, dnnl_graph_partition_policy_t policy);
+```
+
+```cpp
+class graph {
+    ...
+    std::vector<partition> get_partitions(
+            partition::policy policy = partition::policy::fusion);
+    ...
+};
+```
+
+A partition policy is used in the API to control granularities
+of optimizations to be applied when partitioning.
+Currently, oneDNN Graph API supports two partition policies:
+
+```c
+typedef enum {
+    dnnl_graph_partition_policy_fusion = 1,
+    dnnl_graph_partition_policy_debug = 2,
+} dnnl_graph_partition_policy_t;
+```
+
+```cpp
+class partition {
+    ...
+    enum class policy {
+        fusion = dnnl_graph_partition_policy_fusion,
+        debug = dnnl_graph_partition_policy_debug,
+    };
+    ...
+};
+```
+
+- Fusion policy (default partition policy): fusion policy returns partitions
+  with [pre-defined fusion patterns](https://oneapi-src.github.io/oneDNN/dev_guide_graph_fusion_patterns.html),
+  e.g. Convolution followed by a chain of post-ops, MHA, MLP, etc.
+- Debug policy: debug policy returns partitions with single operation in each partition.
+  This policy doesn't apply any fusions, which is useful when
+  user wants to debug or compare performance with fusion policy.
+
+## Proposal
+
+The API to expose max policy to users is similar to fusion and debug policies.
+
+```c
+typedef enum {
+    dnnl_graph_partition_policy_max = 0, // the new max policy
+    dnnl_graph_partition_policy_fusion = 1,
+    dnnl_graph_partition_policy_debug = 2,
+} dnnl_graph_partition_policy_t;
+```
+
+```cpp
+class partition {
+    ...
+    enum class policy {
+        max = dnnl_graph_partition_policy_max, // the new max policy
+        fusion = dnnl_graph_partition_policy_fusion,
+        debug = dnnl_graph_partition_policy_debug,
+    };
+    ...
+};
+```
+
+The expected behaviors when using max policy:
+
+- If not specified, the default partition policy remains to be fusion policy.
+- Max policy returns partitions max in partition size. The criteria of partitioning
+  is based on pre-defined ops list and pre-defined rules. Connected ops
+  that meet the conditions of pre-defined ops list and pre-defined rules
```

Comment:
I think this sentence may be ambiguous, `meet the condistions of pre-defined ops list and don't hit the pre-defined excluding rules will be selected into one partition` ?

Reply:
`meet the conditions of pre-defined ops list and pre-defined rules will be selected into one partition` 
means both pre-defined ops list and pre-defined rules need to be met to be selected into one partition, violate either will not be selected into one partition.

Reply:
Yes, but i remember the rules are negative list. If op met one of rules, the op will be remove from the partition

Reply:
Oh I got your point. But since the pre-defined rules are backend specific, we cannot say it's always a negative list. So here I just keep general words.

Reply:
Will the pre-defined ops list and pre-defined rules be expose through API or API documentation? If not, how does the user know when to use the max policy? Or are you suggesting users to always specify max policy for their models?

Reply:
> Will the pre-defined ops list and pre-defined rules be expose through API or API documentation?

Fusion policy doesn't expose fusion patterns through API, so I think max policy follows similar strategy. For document, maybe we can provide a page like [pre-defined fusion patterns](https://oneapi-src.github.io/oneDNN/dev_guide_graph_fusion_patterns.html), if it helps.

> If not, how does the user know when to use the max policy? Or are you suggesting users to always specify max policy for their models?

I feel this discussion can be merged with https://github.com/oneapi-src/oneDNN/pull/1694/files#r1283768894